rumqttd/src/link/persistance.rs

```diff
@@ -0,0 +1,325 @@
+use crate::link::local::{LinkError, LinkRx, LinkTx};
+use crate::link::network;
+use crate::link::network::Network;
+use crate::local::Link;
+use crate::protocol::{AsyncProtocol, Connect, LastWill, Packet, self};
+use crate::router::{Event, Notification};
+use crate::{ConnectionId, ConnectionSettings};
+
+use disk::Storage;
+use flume::{Receiver, RecvError, SendError, Sender, TrySendError};
+use tracing::{info, error, trace};
+use std::collections::{VecDeque};
+use std::{io, fs};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::error::Elapsed;
+use tokio::{select, time};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("I/O")]
+    Io(#[from] io::Error),
+    #[error("Zero keep alive")]
+    ZeroKeepAlive,
+    #[error("Not connect packet")]
+    NotConnectPacket(Packet),
+    #[error("Network {0}")]
+    Network(#[from] network::Error),
+    #[error("Timeout")]
+    Timeout(#[from] Elapsed),
+    #[error("Channel send error")]
+    Send(#[from] SendError<(ConnectionId, Event)>),
+    #[error("Channel recv error")]
+    Recv(#[from] RecvError),
+    #[error("Persistent session requires valid client id")]
+    InvalidClientId,
+    #[error("Unexpected router message")]
+    NotConnectionAck,
+    #[error("ConnAck error {0}")]
+    ConnectionAck(String),
+    #[error("Channel try send error")]
+    TrySend(#[from] TrySendError<(ConnectionId, Event)>),
+    #[error("Link error = {0}")]
+    Link(#[from] LinkError),
+}
+
+/// Orchestrates between Router and Network.
+pub struct PersistanceLink<P: AsyncProtocol> {
+    _connect: Connect,
+    pub(crate) client_id: String,
+    pub(crate) connection_id: ConnectionId,
+    pub(crate) network: Network<P>,
+    link_tx: LinkTx,
+    link_rx: LinkRx,
+    notifications: VecDeque<Notification>,
+    disk_handler: DiskHandler<P>,
+    network_update_rx: Receiver<Network<P>>,
+    connack: Notification,
+}
+
+pub(super) struct DiskHandler<P: AsyncProtocol> {
+    storage: Storage,
+    protocol: P,
+}
+
+impl<P: AsyncProtocol> DiskHandler<P> {
+    fn new(client_id: &str, protocol: P) -> Result<Self, Error> {
+        // TODO: take params from config
+        let path = format!("/tmp/rumqttd/{}", &client_id);
```

Comment:
to get temporary directory, we should use
```rust 
env::temp_dir();
```
  - [docs](https://doc.rust-lang.org/std/env/fn.temp_dir.html)

Reply:
> On Unix, returns the value of the TMPDIR environment variable if it is set
ideally i'd want this directory to be not dependent on an environment variable since on broker restart I want to reliably find the directory we previously wrote to